## Coding Basics for Researchers - Day 1

*Notebook by [Pedro V Hernandez Serrano](https://github.com/pedrohserrano)*

---
# 2. Pandas Fundamentals
* [2.1. Dealing with different data sources](#2.1)
* [2.2. Data structures](#2.2)

---

![](../data/assam_1.png)

## A research use case

#### Insurgency-Civilian Relations & EU Policy

- International Relations, Peace and Policy Studies
- PhD proposal, S. Roerigh 2020
- Understand patterns in international relations and make effective policies and strategic decisions.
- Normally: 
    - is studied the social order, and
    - lived experiences of civilians living under rebel rule
- Proposed:
    - how and when such factors are perceived, discussed and considered in third party policy decisions
    - what impact third party policy decisions have on rebel-civilian relations. 
- It contributes to a critical perspective and discussion on the motivations, interests, effectiveness of third party policymakers

---
## 2.1. Dealing with different data sources
<a id="2.1">

* Pandas is a widely-used Python library for handling data, particularly on tabular data.
* Borrows many features from R's dataframes:
  - Two-dimensional table whose columns have names and potentially have different types of data types/
* Load it with `import pandas as pd`. The alias pd is commonly used for Pandas.
* Pandas can handle virtually any kind of formats

![](https://pandas.pydata.org/docs/_images/02_io_readwrite.svg)

In [1]:
import pandas as pd

In [2]:
path = '../data/'

* The columns in a dataframe are the observed variables, and the rows are the observations.
* Pandas uses backslash `\` to show wrapped lines when output is too wide to fit the screen.

**File Not Found:**

> Our lessons store their data files in a `data` sub-directory, which is why the path to the file is `data/...csv`. If you forget to include `data/` or if you have it but your cope of the file is somewhere else, you will bet a runtime error that ends with a line like this:

`ERROR`: _OSError: File b'gapminder_gdp_oceania.csv' does not exist_

#### Dataset 1: Reputation of Terror Groups (RTG) Dataset
Description: The dataset contains all domestic terrorist groups, defined in Enders et al. (2011) and based on the Global Terrorism Database, with more than 5 terrorist attacks from 1980 to 2011. The data is in group name - year format—the data codes terrorist groups' actions which can build a reputation among constituency and out-group. Researchers can find originally coded variables regarding positive and negative reputation among the audience and existing group-level variables.

[Link to data](http://www.efetokdemir.com/data.html)

In [3]:
# READING A STATA FILE
rtg_table = pd.read_stata(path+'replicationdatajpr-oldstata.dta')

In [4]:
rtg_table.head()

year                                              gname  ffund  childrec  \
0  1989                                        1 May Group      0         0   
1  1991                                        1 May Group      0         0   
2  1992                                        1 May Group      0         0   
3  1989  16 January Organization for the Liberation of ...      0         0   
4  1983                                      2 April Group      0         0   

   frec  rebel  parterr  terpwing  teraff  govcaus  ...  nat  civcausreal  \
0     0      0        0         0       0        1  ...  0.0     0.250000   
1     0      0        0         0       0        0  ...  0.0     2.333333   
2     0      0        0         0       0        0  ...  0.0     0.000000   
3     0      0        0         0       0        1  ...  NaN    22.625000   
4     0      0        0         0       0        0  ...  0.0     0.000000   

   civcauseffreal  outnegrep  cleavage  reputation  last  counter  endedtype  \
0             1.0        1.0       1.0         0.0   3.0      1.0        0.0   
1             0.0        0.0       1.0         0.0   3.0      2.0        0.0   
2             0.0        0.0       1.0         0.0   3.0      3.0        1.0   
3             0.0        0.0       NaN         0.0   1.0      1.0        1.0   
4             0.0        1.0       1.0         0.0   1.0      1.0        1.0   

   endedtype2  
0         0.0  
1         0.0  
2         1.0  
3         1.0  
4         1.0  

[5 rows x 50 columns]

#### Dataset 2: The Foundations of Rebel Group Emergence (FORGE) Dataset

Description: It provides information on the origins of violent non-state actors engaged in armed conflict against their government, resulting in 25+ yearly battle deaths, active between 1946 and 2011. The unit of observation in this dataset is the rebel group organization. We also include information on the dyad and conflict in which these groups are participants for easy integration with various Uppsala Conflict Data Program (UCDP) datasets. We draw upon the population of groups included in the Non-State Actor database described in greater detail here:
    
[Link to data](http://ksgleditsch.com/eacd.html)

In [5]:
# READING AN ASCII FILE ON TABULAR FORMAT WITH CSV FUNCTION
forge_table = pd.read_csv(path+'nsa_v3.4_21November2013.asc', delimiter='\t')

In [6]:
forge_table.head()

obsid  ucdpid  dyadid   side_a  acr                          side_b  \
0   NSA.3.4-1       1     462  Bolivia  BOL  Popular Revolutionary Movement   
1   NSA.3.4-4       1     463  Bolivia  BOL                             MNR   
2   NSA.3.4-7       1     464  Bolivia  BOL                             ELN   
3  NSA.3.4-10       2     654   France  FRN                   Khmer Issarak   
4  NSA.3.4-13       3     466    China  CHN         Peoples Liberation Army   

    startdate     enddate oldid oldconfid  ...  rsupname  gov.support  \
0  1946-06-01  1946-07-21  1010    1010.1  ...       NaN           no   
1  1952-04-09  1952-04-12  1010    1010.2  ...       NaN           no   
2  1967-03-01  1967-10-16  1010    1010.3  ...      Cuba     explicit   
3  1946-08-01  1953-11-09  1020      1020  ...  Thailand     explicit   
4  1946-01-01   1949-10-1  1030      1030  ...      USSR     explicit   

   gtypesup gsupname  govextpart  type.of.termination  victory.side  \
0       NaN      NaN          no                  4.0           2.0   
1       NaN      NaN          no                  4.0           2.0   
2  military      USA          no                  4.0           1.0   
3  military      USA          no                  7.0           NaN   
4  military      USA          no                  4.0           2.0   

  prevactive prevact.ref    oldobsid  
0          0         NaN   NSA.3.3-1  
1          0         NaN   NSA.3.3-4  
2          0         NaN   NSA.3.3-7  
3          0         NaN  NSA.3.3-10  
4          0         NaN  NSA.3.3-13  

[5 rows x 43 columns]

#### Dataset 3: The CEPS EurLex dataset: EU laws from 1952-2019 with full text and 22 variables

Description: The dataset contains 142.036 EU laws - almost the entire corpus of the EU's digitally available legal acts passed between 1952 - 2019. It encompasses the three types of legally binding acts passed by the EU institutions: 102.304 regulations, 4.070 directives, 35.798 decisions in the English language. The dataset was scraped from the official EU legal database (Eur-lex.eu) and transformed in machine-readable CSV format with R and Python programming languages.   
The Centre collected the European Policy Studies (CEPS) dataset for the TRIGGER project (https://trigger-project.eu/). We hope that it will facilitate future quantitative and computational research on the EU. 

[Link to data](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/0EGYWY)

In [7]:
# READING AN EXCEL FILE
eurlex_table = pd.read_excel(path+'EurLex_all_no_text.xlsx')

In [8]:
eurlex_table.head()

CELEX                                           Act_name  Act_type  \
0  32019D0276  Decision (EU) 2019/276 of the European Parliam...  Decision   
1  32019D0277  Decision (EU) 2019/277 of the European Parliam...  Decision   
2  32019D0275  Decision (EU) 2019/275 of the European Parliam...  Decision   
3  32018D1859  Decision (EU) 2018/1859 of the European Parlia...  Decision   
4  32018D1720  Decision (EU) 2018/1720 of the European Parlia...  Decision   

     Status                                            EUROVOC  \
0  In Force  aid to refugees; budget appropriation; EC gene...   
1  In Force  aid to catastrophe victims; emergency aid; EC ...   
2  In Force  professional reintegration; Attica; EGF; EC ge...   
3  In Force  commitment appropriation; Latvia; payment appr...   
4  In Force  Northern Portugal; Portugal; employment aid; e...   

                                      Subject_matter Treaty  \
0  cooperation policy;  budget;  EU finance;  int...   TFEU   
1  cooperation policy;  EU finance;  budget;  det...   TFEU   
2  employment;  regions of EU Member States;  EU ...   TFEU   
3  budget;  Europe;  EU finance;  cooperation pol...   TFEU   
4  regions of EU Member States;  Europe;  economi...   TFEU   

            Legal_basis_celex                                Authors  \
0              32013Q1220(01)  European Parliament; European Council   
1  32002R2012; 32013Q1220(01)  European Parliament; European Council   
2  32013Q1220(01); 32013R1309  European Parliament; European Council   
3  32002R2012; 32013Q1220(01)  European Council; European Parliament   
4  32013Q1220(01); 32013R1309  European Council; European Parliament   

  Procedure_number  ... Temporal_status              Act_cites  \
0              NaN  ...             NaN             32013R1311   
1              NaN  ...             NaN             32013R1311   
2              NaN  ...             NaN             32013R1311   
3              NaN  ...             NaN  32018D508; 32013R1311   
4              NaN  ...             NaN             32013R1311   

                                         Cites_links Act_ammends  \
0         http://data.europa.eu/eli/reg/2013/1311/oj         NaN   
1         http://data.europa.eu/eli/reg/2013/1311/oj         NaN   
2         http://data.europa.eu/eli/reg/2013/1311/oj         NaN   
3  http://data.europa.eu/eli/dec/2018/508/oj; htt...         NaN   
4         http://data.europa.eu/eli/reg/2013/1311/oj         NaN   

  Ammends_links Eurlex_link ELI_link Proposal_link  Oeil_link  Additional_info  
0           NaN         NaN      NaN           NaN        NaN              NaN  
1           NaN         NaN      NaN           NaN        NaN              NaN  
2           NaN         NaN      NaN           NaN        NaN              NaN  
3           NaN         NaN      NaN           NaN        NaN              NaN  
4           NaN         NaN      NaN           NaN        NaN              NaN  

[5 rows x 23 columns]

---
## 2.2. Data structures
<a id="2.2">

A DataFrame is a collection of Series; The DataFrame is how Pandas represents a table, and Series is the data structure Pandas use to represent a column.

Pandas is built on top of the Numpy library, which means that most of the methods defined for Numpy Arrays apply to Pandas Series/DataFrames.

What makes Pandas so attractive is the powerful interface to access individual table records, proper handling of missing values, and relational-databases operations between DataFrames.

In [9]:
# Basic representation of a dataframe
dictionary = {'age': [20, 30], 'height': [1.80, 1.60], 'course':['Python', None]}

# Define a dataframe
df = pd.DataFrame(data=dictionary)

# Print the dataframe
df

age  height  course
0   20     1.8  Python
1   30     1.6    None

Use the `DataFrame.info()` method to find out more about a dataframe

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     2 non-null      int64  
 1   height  2 non-null      float64
 2   course  1 non-null      object 
dtypes: float64(1), int64(1), object(1)
memory usage: 176.0+ bytes


* This is a DataFrame
* Columns named `age`, `height` and `course`
* Two actual 64-bit int values and one is floating point
* The columns are not null
* Uses 176 bytes of memory

The DataFrame.columns variable stores information about the dataframe's columns

* Note that this is data, _not_ a method. (it doesn't have parentheses)
  - Like `math.pi`
  - So do not use `()` to call it

In [11]:
df.columns

Index(['age', 'height', 'course'], dtype='object')

Use `DataFrame.T` to transpose a dataframe:

* Sometimes want to treat columns as rows and vice versa.
* Transpose (written `.T`) doesn't copy the data, just changes the program's view of it.
* Like `columns`, it is a member variable


In [12]:
df.T

0     1
age         20    30
height     1.8   1.6
course  Python  None

Use `DataFrame.describe()` to get summary statistics about data

`DataFrame.describe()` gets the summary statistics of only the columns that have numerical data. All other columns are ignored, unless you use the argument `include = 'all'` 

In [13]:
df.describe()

age    height
count   2.000000  2.000000
mean   25.000000  1.700000
std     7.071068  0.141421
min    20.000000  1.600000
25%    22.500000  1.650000
50%    25.000000  1.700000
75%    27.500000  1.750000
max    30.000000  1.800000

Make use of a list to select desired columns and create a subset

In [14]:
# Desired columns
my_columns = ['age','course']

# Selection of subset
df[my_columns]

age  course
0   20  Python
1   30    None

Use `DataFrame.to_csv()` to generate a CSV file as result of the analysed dataframe

In [15]:
df[my_columns].to_csv('example.csv', index=False)

## EXERCISES

+ _1. Reproduce the examples given in the introduction of this notebook

    - Download 1 of the three datasets from section 2.1
    - Use Pandas to 
        - define a path to the files from your local computer,
        - read it using a variant of the `read` function and 
        - save it in a Python object 
___

+ _2.  Use the `help()` function to analyse Python objects and methods 
    - What is the difference between the object `df.columns` and `df.index`?
    - What is the difference between the methods `df.head` and `df.tail`?

Can you tell the difference between an object and a method?

**Note:** you must substitute `df` for the name of your dataframe

---


+ _3.  Use the `df.info()` and `df.describe()`functions to learn general information about the datasets

- What is the information presented using `df.info()`?
- What is the information presented using `df.describe()`?

**Note:** you must substitute `df` for the name of your dataframe

---


+ _4.  Using the `CEPS EurLex dataset`.

    - Get the name of the columns using `df.columns`
    - Create a subset of the original dataset by only selecting the columns called `Act_name` and `Eurlex_link`
    - Using your newly created subset, sort the dataset alphabetically based on the column `Act_name`
    - Save the sorted subset in a `CSV` file
    - Check your filtered data. Is it what we expected?

---
